cluster-autoscaler/cloudprovider/clusterapi/clusterapi_converters_test.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clusterapi
+
+import (
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+const (
+	apiVersion = "machine.openshift.io/v1beta1"
+	kind       = "MachineSet"
+	name       = "Foo"
+	namespace  = "Bar"
+)
+
+func TestConverterNewMachineSetFromUnstructured(t *testing.T) {
+	testobj := unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": apiVersion,
+			"kind":       kind,
+			"metadata": map[string]interface{}{
+				"name":      name,
+				"namespace": namespace,
+			},
+			"spec": map[string]interface{}{
+				"template": map[string]interface{}{
+					"spec": map[string]interface{}{
+						"metadata": map[string]interface{}{
+							"labels": map[string]interface{}{
+								"foo": "bar",
+							},
+						},
+						"taints": []interface{}{
+							map[string]interface{}{
+								"key":       "taintkey",
+								"value":     "taintvalue",
+								"effect":    "tainteffect",
+								"timeadded": metav1.Time{},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
```

Comment:
Does this cover all of the fields in a `MachineSet`?

Reply:
not yet

Reply:
so, i have some questions about how deep we should go with these things. as i descend into more and more objects i start to wonder if we should be testing the basic copy stuff?

i am more feeling that we should test the extra bits we have added, eg taint replications and label deep copy, rather than testing to see if the all the fields are just copied. that said, i have left the extra tests i added because it is probably better for coverage to have this stuff.

Reply:
as per our discussions, i am adding more tests here and making them more-or-less a bunch of brute force tests. i have also added a jira ticket (see OCPCLOUD-826) to address improving the tests.